Log if a server certificate lacks the subjectAlternativeName extensions

amcasey opened the pull request:
# Log if a server certificate lacks the subjectAlternativeName extensions

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

## Description

CommonName was [deprecated](https://developer.chrome.com/blog/chrome-58-deprecations/#remove-support-for-commonname-matching-in-certificates) in favor of subjectAlternativeName so there's a good chance of getting a browser security warning if it's missing.

amcasey commented:
Suggested by @JamesNK at the Tuesday meeting.

amcasey commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs:
```diff
@@ -76,7 +76,7 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, MultiplexedConnectionDe
         if (listenOptions.HttpsOptions != null)
         {
-            var sslServerAuthenticationOptions = HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions);
+            var sslServerAuthenticationOptions = HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions, Trace);
```
This is the reason we use `ILogger`, rather than `ILogger<HttpsConnectionMiddleware>`.  Is there a better way?

amcasey replied:
No longer using `ILogger`.

amcasey commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
     [LoggerMessage(8, LogLevel.Debug, "Failed to open certificate store {StoreName}.", EventName = "FailToOpenStore")]
     public static partial void FailedToOpenStore(this ILogger<HttpsConnectionMiddleware> logger, string? storeName, Exception exception);
 
+    [LoggerMessage(9, LogLevel.Information, "Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension and may not be accepted by browsers.", EventName = "NoSubjectAlternativeName")]
```
How unique do IDs have to be?  I would have guessed per-logger-type, but the non-generic `ILogger` extension methods seem to reuse them freely.

JamesNK replied:
Uniqueness is the `ILogger` name +  logger message ID (which is ID and string)

Usually, the logger used by a type has the same name as the type, e.g. `Microsoft.AspNetCore.Kestrel.HttpConnectionMiddleware`. Then all the logs defined here can have their own IDs that start from 1.

tldr as long as they are unique in this file you should be fine.

BrennanConroy replied:
But in this case since you're passing in KestrelTrace you're breaking that assumption. Should probably try to get an `ILogger<HttpsConnectionMiddleware>` in `BindAsync` and pass that in.

JamesNK replied:
Yes. The KestrelTrace logger won't have the name `Microsoft.AspNetCore.Kestrel.HttpConnectionMiddleware`. It has its own `ILogger` name and will already have a log message with ID=9. So that won't be unique.

BrennanConroy replied:
> What would be the consequences of sharing a logger across endpoints?

Shouldn't be any, they are already effectively Singletons in DI.

I vote for whatever option gives us an `ILogger<HttpsConnectionMiddleware>`.

amcasey replied:
@BrennanConroy Tell me more about the expectation that they'll be singletons?  That does seem to be the case for the [default impl](https://source.dot.net/#Microsoft.Extensions.Logging/LoggerFactory.cs,143), but it's not clear from the name ("factory", rather than "provider") or the [doc comment](https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/ILoggerFactory.cs,15)
```csharp
        /// <summary>
        /// Creates a new <see cref="ILogger"/> instance.
        /// </summary>
        /// <param name="categoryName">The category name for messages produced by the logger.</param>
        /// <returns>The <see cref="ILogger"/>.</returns>
        ILogger CreateLogger(string categoryName);
```

BrennanConroy replied:
https://source.dot.net/#Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs,38
and
https://source.dot.net/#Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs,39
is where logging gets added for use in ASP.NET Core apps. And AddLogging is called from the host builders.

If you use `ILoggerFactory` it has a cache of loggers that it creates/gets from:
https://source.dot.net/#Microsoft.Extensions.Logging/LoggerFactory.cs,143 and the `ILogger` impl we use, actually just calls the `ILoggerFactory` impl: https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LoggerT.cs,26

So at the end of the day it's all singleton loggers.

amcasey commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
     public static partial void FailedToOpenStore(this ILogger<HttpsConnectionMiddleware> logger, string? storeName, Exception exception);
 
+    [LoggerMessage(9, LogLevel.Information, "Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension and may not be accepted by browsers.", EventName = "NoSubjectAlternativeName")]
+    public static partial void NoSubjectAlternativeName(this ILogger logger, string thumbprint);
```
This should really be `ILogger<HttpsConnectionMiddleware>`.

JamesNK replied:
It doesn't matter other than consistency.

amcasey commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
     [LoggerMessage(8, LogLevel.Debug, "Failed to open certificate store {StoreName}.", EventName = "FailToOpenStore")]
     public static partial void FailedToOpenStore(this ILogger<HttpsConnectionMiddleware> logger, string? storeName, Exception exception);
 
+    [LoggerMessage(9, LogLevel.Information, "Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension and may not be accepted by browsers.", EventName = "NoSubjectAlternativeName")]
```
I went with `Information`, rather than `Warning`, because this server might not receive auth requests from browsers (i.e. it might not actually matter that the extension is missing.).

JamesNK replied:
Info is fine. I think we need to provide more detail about why browsers might not accept it. Is there a good doc or article that we could link to?

> Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension. Some browsers require a SAN and will not accept this certificate. For more information, see https://aka.ms/aspnet/cert-san.

amcasey replied:
I linked an article in the description, but we don't own it.

amcasey replied:
I found this, but it seems to be about running a CA (and seems nervous about using SANs at all?).
https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2008-R2-and-2008/ff625722(v=ws.10)?redirectedfrom=MSDN#BKMK_Security

amcasey replied:
@vcsjones Is this (i.e. specifying a subject alternative name) even a sensible thing to warn about / recommend?  Maybe we should only do it for self-signed certs or when the common name is `localhost`?

amcasey replied:
Thanks!

blowdart replied:
Even our internal cert issues implemented SANs 10 years+ ago (Because muggins here wrote the request UI).

But whilst the cab forum rules are meant to be "Thou shalt not" there are moments when CAs forget themselves, due to bugs, so a information message is still a good idea.

amcasey replied:
I don't suppose either @vcsjones or @blowdart is aware of a page I could link to?  (There are some useful pages online, but I understand aka.ms links must point to 1P properties.)

blowdart replied:
I don't know of anything 1st party

amcasey reviewed (commented):
Some notes on the parts I don't like.  Suggestions?

blowdart commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -452,12 +452,16 @@ private static async ValueTask<SslServerAuthenticationOptions> ServerOptionsCall
         {
             throw new InvalidOperationException(CoreStrings.FormatInvalidServerCertificateEku(certificate.Thumbprint));
         }
+        else if (!CertificateLoader.DoesCertificateHaveASubjectAlternativeName(certificate))
```
You could go further here. In modern browsers the SAN is selected by default if present and the subject is ignored. If you have a SAN that doesn't contain the subject that may also be log worthy.

amcasey replied:
I think the suggestion here was to log in cases where the SAN doesn't incorporate the CommonName.  Assuming I understood that correctly, are there helpers for checking that?  There seem to be quite a few different ways to specify SANs, including wildcards, and I'm not sure I'd want to write a validator from scratch.  But maybe there's just a helper for "does this name match that SAN"?

blowdart replied:
I've not see anything, and it's fair if you want to avoid it because it may lead to false positives.

amcasey committed:
Log if a server certificate lacks the subjectAlternativeName extensions

CommonName was deprecated in favor of subjectAlternativeName so there's a good chance of getting a browser security warning if it's missing.

JamesNK commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
     public static partial void FailedToOpenStore(this ILogger<HttpsConnectionMiddleware> logger, string? storeName, Exception exception);
 
+    [LoggerMessage(9, LogLevel.Information, "Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension and may not be accepted by browsers.", EventName = "NoSubjectAlternativeName")]
+    public static partial void NoSubjectAlternativeName(this ILogger logger, string thumbprint);
```
Exact logger type now works. Be consistent with other log methods:
```suggestion
    public static partial void NoSubjectAlternativeName(this ILogger<HttpsConnectionMiddleware> logger, string thumbprint);
```

amcasey replied:
Just a bad copy-paste on my part.

amcasey commented on src/Servers/Kestrel/Core/src/CertificateLoader.cs:
```diff
@@ -106,6 +106,9 @@ internal static bool IsCertificateAllowedForServerAuth(X509Certificate2 certific
         => certificate.HasPrivateKey;
 
+    internal static bool DoesCertificateHaveASubjectAlternativeName(X509Certificate2 certificate)
+        => certificate.Extensions.OfType<X509SubjectAlternativeNameExtension>().Any();
```
I'm also not used to using `OfType` - I copied it from `IsCertificateAllowedForServerAuth`.

I'm probably missing something, but it looks like [the indexer](https://source.dot.net/#System.Security.Cryptography/System/Security/Cryptography/X509Certificates/X509ExtensionCollection.cs,56) does the same linear search?  And it doesn't appear to use the struct enumerator pattern, so writing the foreach loop myself probably wouldn't be a win either?

JamesNK replied:
The foreach in the indexer is over a `List<X509Extension>`, so will use its struct enumerator.

The indexer looks like it allocates an Oid. There are allocations no matter what we do.

I guess just leave it as is.

amcasey commented on src/Servers/Kestrel/Core/src/KestrelServer.cs:
```diff
@@ -35,7 +35,7 @@ public KestrelServer(IOptions<KestrelServerOptions> options, IConnectionListener
             new[] { transportFactory ?? throw new ArgumentNullException(nameof(transportFactory)) },
             Array.Empty<IMultiplexedConnectionListenerFactory>(),
-            new SimpleHttpsConfigurationService(),
+            new SimpleHttpsConfigurationService(loggerFactory.CreateLogger<HttpsConnectionMiddleware>()),
```
Does that have an advantage beyond being idiomatic?

amcasey replied:
Done but I'm still curious about the reasoning.

JamesNK replied:
Usually, either types create their own logger with `ILoggerFactory`, or DI does it for them. Creating a type explicitly and making the creator create the logger feels off.

I don't think there is explicit reason except maybe it's about ownership. `SimpleHttpsConfigurationService` is using the logger so it should create it. IDK 🤷‍♂️

amcasey committed:
Address minor PR feedback

amcasey commented:
@BrennanConroy Any remaining concerns?
